src/controller/CHIPDeviceController.cpp

```diff
@@ -2189,46 +2194,166 @@ void DeviceCommissioner::OnDeviceConnectionRetryFn(void * context, const ScopedN
 }
 #endif // CHIP_DEVICE_CONFIG_ENABLE_AUTOMATIC_CASE_RETRIES
 
-// ClusterStateCache::Callback impl
+// ClusterStateCache::Callback / ReadClient::Callback
 void DeviceCommissioner::OnDone(app::ReadClient * readClient)
 {
     VerifyOrDie(readClient != nullptr && readClient == mReadClient.get());
     mReadClient.reset();
     switch (mCommissioningStage)
     {
     case CommissioningStage::kReadCommissioningInfo:
-        // Silently complete the stage, data will be saved in attribute cache and
-        // will be parsed after all ReadCommissioningInfo stages are completed.
-        CommissioningStageComplete(CHIP_NO_ERROR);
-        break;
-    case CommissioningStage::kReadCommissioningInfo2:
-        // Note: Only parse commissioning info in the last ReadCommissioningInfo stage.
-        ParseCommissioningInfo();
+        ContinueReadingCommissioningInfo(mCommissioningDelegate->GetCommissioningParameters());
         break;
     default:
         VerifyOrDie(false);
         break;
     }
 }
 
-void DeviceCommissioner::ParseCommissioningInfo()
+namespace {
+// Helper for grouping attribute paths into read interactions
+class ReadInteractionBuilder
 {
-    CHIP_ERROR err = CHIP_NO_ERROR;
-    ReadCommissioningInfo info;
+    static constexpr auto kCapacity = InteractionModelEngine::kMinSupportedPathsPerReadRequest;
 
-    err = ParseCommissioningInfo1(info);
-    if (err == CHIP_NO_ERROR)
+    size_t mSkip  = 0;
+    size_t mCount = 0;
+    app::AttributePathParams mPaths[kCapacity];
+
+public:
+    ReadInteractionBuilder(size_t skip = 0) : mSkip(skip) {}
+
+    size_t size() { return std::min(mCount, kCapacity); }
+    bool exceeded() { return mCount > kCapacity; }
+    app::AttributePathParams * paths() { return mPaths; }
+
+    // Adds an attribute path if within the current window.
+    // Returns false if the available space has been exceeded.
+    template <typename... Ts>
+    bool AddAttributePath(Ts &&... args)
     {
-        err = ParseCommissioningInfo2(info);
+        VerifyOrReturnValue(mSkip == 0, true, mSkip--);
+        VerifyOrReturnValue(mCount < kCapacity, false, mCount = kCapacity + 1);
```

Comment:
These should be if's . The 3 complex args are really really hard to parse on a single read.

Reply:
Done